### PR TITLE
feat(tuner): onnx model dump load 280

### PR DESCRIPTION
**Pull request by gvondulong**
This pr adds and test export to onnx from pytorch, keras and paddle

### CODE REVIEW

File: finetuner/tuner/onnx.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+)
+
+from finetuner.helper import get_framework
+
+
+def to_onnx(
+    embed_model: 'AnyDNN',
+    path: str,
+    input_shape: Union[Tuple[int], List[int]],
+    batch_size: Optional[Union[int, None]] = None,
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:19]: I think `batch_size` should not be an argument here. It's an extra parameter for the user to consider and it doesn't offer anything

File: finetuner/tuner/onnx.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    opset_version: int = 11,
+) -> None:
+    """Func that converts a given model in paddle, torch or keras,
+    and converts it to the ONNX format
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:23]: syntax

File: tests/unit/tuner/test_onnx.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    session = onnxruntime.InferenceSession(temp_onnx_file)
+
+    model.eval()
+    y_original = model(torch.Tensor(x)).detach().numpy()
+    y_exported = session.run(None, {session.get_inputs()[0].name: x})[0]
+
+    np.testing.assert_allclose(y_original, y_exported, rtol=1e-03, atol=1e-05)
+
+
+def test_pytorch_lstm_to_onnx(get_pytorch_lstm_model):
```
Comment (gvondulong) [inline tests/unit/tuner/test_onnx.py:203]: @gmastrapas, contrary to what we spoke about, concerning putting the onnx evaluation into the onnx.py file, I put it into the test file. 
Even though it would be nice to have such functionality, as a user, the different types of models e.g. LSTM have outputs different from tensors. In case of lstm, it outputs the hidden state and the memory vector. 
We would need to handle all output types for all possible types of models. 
IMHO, that would be a little over-engineered and in the end it is only a few lines of code to evaluate the export, as a user.

File: finetuner/tuner/onnx.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    :param input_shape: Input shape of embedding model
+    :param batch_size: The batch size the model was trained with. This
+      is only required when converting pyTorch model to onnx. In other
+      frameworks, the batch size does not need to be made specific
+    :param opset_version: ONNX opset version in which to register
+    """
+    if isinstance(input_shape, tuple):
+        input_shape = list(input_shape)
+
+    if not path.endswith(".onnx"):
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:35]: use single quotes

File: finetuner/tuner/onnx.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        return {
+            'torch': _to_onnx_torch,
+            'keras': _to_onnx_keras,
+            'paddle': _to_onnx_paddle,
+        }[framework_name](embed_model, path, input_shape, opset_version)
+
+    fm = get_framework(embed_model)
+
+    # Get framework-specific func to register model in ONNX
+    _parse_and_apply_to_onnx_func(fm)
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:49]: why not `getattr(globals())`. It's a single line

File: finetuner/tuner/onnx.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    :param embed_model: Embedding model to register in ONNX
+    :param path: Patch where to register ONNX model to
+    :param input_shape: Embedding model input shape
+    :param batch_size: The batch size the model was trained with
+    :param opset_version: ONNX opset version in which to register
+    """
+
+    import torch
+
+    embed_model.eval()
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:81]: if model is previously in training mode, it should stay that way

File: requirements.txt [hunk 7]
```diff
@@ -1 +1,4 @@
-.
\ No newline at end of file
+onnx==1.9.0
```
Comment (gmastrapas) [inline requirements.txt:1]: I don't think we should add these in requirements.txt. We should only try and import them in the functions that they are needed and raise errors if not found. e.g.
```
The onnx package is required to export a model in ONNX. Try 'pip install onnx'
```

File: tests/unit/tuner/test_onnx.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+from pathlib import Path
+from typing import Any, Tuple
+
+from tensorflow import keras
+
+from finetuner.helper import get_framework
+from finetuner.tuner.onnx import to_onnx
+
+
+def validate_onnx_export(
```
Comment (gmastrapas) [inline tests/unit/tuner/test_onnx.py:18]: I really think this should be in our public API

File: tests/unit/tuner/test_onnx.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+    if is_training_before:
+        embed_model.train()
+
+    np.testing.assert_allclose(y_original, y_exported, rtol=1e-03, atol=1e-05)
+
+
+@pytest.fixture
+def get_pytorch_linear_model():
+    def _get_model(input_shape):
```
Comment (gmastrapas) [inline tests/unit/tuner/test_onnx.py:78]: why not a single function? non pytest.fixture

File: finetuner/tuner/onnx.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    Union,
+)
+
+from finetuner.helper import get_framework
+
+
+def to_onnx(
+    embed_model: 'AnyDNN',
+    path: str,
+    input_shape: Union[Tuple[int], List[int]],
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:18]: `Sequence[int]`?

File: finetuner/tuner/onnx.py (hunk 1 above)
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:19]: `Optional[int]`

File: finetuner/tuner/onnx.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+) -> None:
+    """Convert a paddle embedding model to the ONNX format
+    :param embed_model: Embedding model to register in ONNX
+    :param path: Patch where to register ONNX model to
+    :param input_shape: Embedding model input shape
+    :param opset_version: ONNX opset version in which to register
+    """
+
+    # Removing onnx extension as paddle adds it automatically
+    if path.endswith(".onnx"):
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:144]: single quotes

**Review by gmastrapas — CHANGES_REQUESTED** [review]
Left some comments

File: tests/unit/tuner/test_onnx.py (hunk 3 above)
Comment (gmastrapas) [reply tests/unit/tuner/test_onnx.py:203]: I think we can stick to embedding models, i.e. assume that all models return a batch of vectors `(batch_size, dim)`

File: finetuner/tuner/onnx.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+import numpy as np
+import onnxruntime
+
+from finetuner.helper import AnyDNN
+from finetuner.helper import get_framework
+
+
+def to_onnx(
+    embed_model: AnyDNN,
+    path: str,
```
Comment (bwanglzu) [inline finetuner/tuner/onnx.py:11]: i suggest we use `model_path` here, consistent with what we have in `def save(..)`. (see tuner/__init__.py)

File: finetuner/tuner/onnx.py (hunk 2 above)
Comment (gvondulong) [reply finetuner/tuner/onnx.py:16]: Can you elaborate?

File: finetuner/tuner/onnx.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    :param input_shape: Embedding model input shape
+    :param batch_size: The batch size during export
+    :param opset_version: ONNX opset version in which to register
+    """
+
+    import torch
+
+    embed_model.eval()
+
+    x = torch.randn([batch_size] + input_shape, requires_grad=True, dtype=torch.float32)
```
Comment (bwanglzu) [inline finetuner/tuner/onnx.py:73]: why we stick with `float32` as dtype? does it work for different kind of models? we have `input_dtype` as `float32` as default in tailor, i think we need to have some flexibility here. see [this](https://github.com/jina-ai/finetuner/blob/main/finetuner/tailor/base.py#L13) link

File: finetuner/tuner/onnx.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+    # Removing onnx extension as paddle adds it automatically
+    if path.endswith(".onnx"):
+        path = path[:-5]
+
+    import paddle
+    from paddle.static import InputSpec
+
+    shape = [None] + list(input_shape)
+    x_spec = InputSpec(shape, 'float32', 'input')
```
Comment (bwanglzu) [inline finetuner/tuner/onnx.py:141]: same, the `input_dtype` should be configurable

File: finetuner/tuner/onnx.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+) -> None:
+    """
+    Test an exported model by comparing the outputs of the original and the exported model
+    against the same input.
+    :param embed_model: The original embedding model. Can be either a PyTorch module,
+        a Keras model or a PaddlePaddle layer.
+    :param export_path: The path where the exported model is stored.
+    :param input_shape: The model's expected input shape, without the batch axis.
+    """
+    fm = get_framework(embed_model)
```
Comment (bwanglzu) [inline finetuner/tuner/onnx.py:165]: i think you can import `onxxruntime` here

File: finetuner/tuner/onnx.py (hunk 2 above)
Comment (gmastrapas) [reply finetuner/tuner/onnx.py:16]:
```
"""Convert a PyTorch, Keras or PaddlePaddle embedding model to the ONNX format"""
```

**Review by bwanglzu — CHANGES_REQUESTED** [review]
some comments

**Comment by bwanglzu** [comment]
can we also make sure some model trained on gpu can be exported correctly as well?

File: tests/unit/tuner/test_onnx.py (hunk 3 above)
Comment (gvondulong) [reply tests/unit/tuner/test_onnx.py:141]: Solved in offline conversation. 
The evaluation is move back as George suggested, since we agreed it would be save to assume, that the embed model creates only one tensor as embedding for new.

File: finetuner/tuner/onnx.py (hunk 13 above)
Comment (tadejsv) [reply finetuner/tuner/onnx.py:73]: But in tailor you don't even use that property, or am i wrong?

File: finetuner/tuner/onnx.py (hunk 12 above)
Comment (gvondulong) [reply finetuner/tuner/onnx.py:11]: Good point. I changed it.

File: finetuner/tuner/onnx.py (hunk 5 above)
Comment (tadejsv) [reply finetuner/tuner/onnx.py:39]: Why not just `if/elif`

File: requirements.txt (hunk 7 above)
Comment (tadejsv) [reply requirements.txt:1]: I agree

File: finetuner/tuner/onnx.py (hunk 14 above)
Comment (gvondulong) [reply finetuner/tuner/onnx.py]: Changed it here.

File: finetuner/tuner/onnx.py (hunk 5 above)
Comment (gvondulong) [reply finetuner/tuner/onnx.py:40]: I changed the approach using `globals`, because I like to make it more explicit. In the end, I am super agnostic here. globals, if/else, my way are all fine to me. It all kinda is the same. :) 
I will change it to if/else, since it might be the most basic and easy way of doing it. 
If you have objections, feel free to raise them. Otherwise I am going to wrap this up?!

File: finetuner/tuner/onnx.py (hunk 13 above)
Comment (bwanglzu) [reply finetuner/tuner/onnx.py:76]: we use, actually for most of the models `float32` works, lstm seems does not (with an `Embedding` layer).

File: tests/unit/tuner/test_onnx.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+def get_paddle_lstm_model():
+    def _get_model():
+        return paddle.nn.Sequential(paddle.nn.LSTM(32, 2))
+
+    return _get_model
+
+
+def test_pytorch_to_onnx(get_pytorch_linear_model):
+    INPUT_SHAPE = [32]
+    temp_onnx_file = str(Path(tempfile.tempdir) / 'finetuned.onnx')
```
Comment (bwanglzu) [inline tests/unit/tuner/test_onnx.py:108]: there is a pytest fixture for this https://docs.pytest.org/en/6.2.x/tmpdir.html

File: tests/unit/tuner/test_onnx.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+def paddle_conv_model():
+    return paddle.nn.Sequential(paddle.nn.Conv2D(3, 3, 2, stride=1))
+
+
+@pytest.fixture
+def get_paddle_lstm_model():
+    def _get_model():
+        return paddle.nn.Sequential(paddle.nn.LSTM(32, 2))
+
+    return _get_model
```
Comment (bwanglzu) [inline tests/unit/tuner/test_onnx.py:103]: for all these fixtures, if you can copy fixtures from `test/unit/tailor/test_torch.py` (and keras, paddle), and move it into `conftest` will be even better. (feel free to not do it, but we need to clean this up)

**Review by bwanglzu — COMMENTED** [review]
some comments in tests

File: tests/unit/tuner/test_onnx.py (hunk 17 above)
Comment (bwanglzu) [reply tests/unit/tuner/test_onnx.py:103]: please ignore this comment, I'll create a common model fixture in projection head PR

**Review by tadejsv — COMMENTED** [review]
Minor comments, looks good in general

File: finetuner/tuner/onnx.py (hunk 13 above)
Comment (JoanFM) [reply finetuner/tuner/onnx.py:76]: I am curious, why is this even needed to export a model.

Comment (JoanFM) [reply finetuner/tuner/onnx.py:76]: It is weird that to export a model I need to pass a batch_size as input

**Review by JoanFM — DISMISSED** [review]

File: requirements.txt (hunk 7 above)
Comment (JoanFM) [reply requirements.txt:1]: yes, makes sense. Try to mimic the ImportExtensions behavior from jina core. But copy paste and adapt, do not add dependency

File: finetuner/tuner/onnx.py (hunk 13 above)
Comment (tadejsv) [reply finetuner/tuner/onnx.py:76]: Creating an example input is needed for pytorch to trace the model for exporting - but we could have a single fixed value for batch size, since it is set as dynamic anyway

File: tests/unit/tuner/test_onnx.py (hunk 16 above)
Comment (gvondulong) [reply tests/unit/tuner/test_onnx.py:108]: Great! Thanks!

File: finetuner/tuner/onnx.py (hunk 13 above)
Comment (JoanFM) [reply finetuner/tuner/onnx.py:76]: at least yes, u shouild have a single squeezed input

Comment (JoanFM) [reply finetuner/tuner/onnx.py:76]: But let's not expose `batch-size`

Comment (gvondulong) [reply finetuner/tuner/onnx.py]: I've added the param

**Review by bwanglzu — APPROVED** [review]
LGTM! Better if you can resolve threads related to `requirements.txt`.

**Comment by bwanglzu** [comment]
@gmastrapas review again?

File: finetuner/tuner/onnx.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def to_onnx(
+    embed_model: AnyDNN,
+    model_path: str,
+    input_shape: Sequence[int],
+    opset_version: int = 11,
+    input_type: str = 'float32',
+) -> None:
+    """Func that converts a given model in PaddlePaddle, PyTorch or Keras,
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:16]: Can we change this docstring?

File: finetuner/tuner/onnx.py [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+        input_spec=[x_spec],
+        opset_version=opset_version,
+    )
+
+
+def validate_onnx_export(
+    embed_model: AnyDNN,
+    export_path: str,
+    input_shape: Tuple[int, ...],
+    batch_size: int = 8,
```
Comment (gmastrapas) [inline finetuner/tuner/onnx.py:176]: Remove batch size from here as well?

File: requirements.txt (hunk 7 above)
Comment (gmastrapas) [reply requirements.txt:1]: will we keep these in the requirements file?

**Review by gmastrapas — CHANGES_REQUESTED** [review]

File: finetuner/tuner/onnx.py (hunk 19 above)
Comment (gvondulong) [reply finetuner/tuner/onnx.py:176]: Would you suggest hard coding it below?

**Review by gmastrapas — CHANGES_REQUESTED** [review]

**Review by gmastrapas — APPROVED** [review]
Looks good 🚀
